### PR TITLE
Add 18ZOO initial configuration

### PR DESCRIPTION
**Pull request by carlorusso**

### CODE REVIEW

**Comment by yzemaze** [comment]
Your svgs aren’t optimized and waste space:
`<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 8 8"><circle cx="4" cy="4" r="4" fill="#ae6d1d"/><text x="4" y="5" text-anchor="middle" font-weight="700" font-size="3" font-family="Lato" fill="#fff">BB</text></svg>`
cf. #3401

**Comment by carlorusso** [comment]
> Your svgs aren’t optimized.
> `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 8 8"><circle cx="4" cy="4" r="4" fill="#ae6d1d"/><text x="4" y="5" text-anchor="middle" font-weight="700" font-size="3" font-family="Lato" fill="#fff">BB</text></svg>`
> cf. #3401

Updated, thanks :)

File: lib/engine/game/g_18_zoo.rb [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+      }.freeze
+
+      MUST_BUY_TRAIN = :always
+
+      # Two lays or one upgrade
+      TILE_LAYS = [{ lay: true, upgrade: true }, { lay: true, upgrade: :not_if_upgraded }].freeze
+
+      CORPORATIONS_ORDER = %w[CR GI PB PE LI TI BB EL] * 2
+
+      TILE_M7 = Tile.from_code('7 (mountain)', 'yellow', 'path=a:0,b:1;upgrade=cost:1,terrain:mountain')
```
Comment (tobymao) [inline lib/engine/game/g_18_zoo.rb:51]: why are these defined like this? why don't you have them in the config?

File: lib/engine/game/g_18_zoo.rb [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+
+      attr_accessor :floated_corporation, :additional_tracks, :new_train_brought, :skip_free_actions
+      attr_reader :companies_for_isr, :companies_for_monday, :market_infos, :available_companies, :future_companies
+
+      def title
+        '18ZOO - Map A (5 families)'
+      end
+
+      def init_optional_rules(optional_rules)
+        optional_rules = super(optional_rules)
```
Comment (tobymao) [inline lib/engine/game/g_18_zoo.rb:74]: just call super

File: lib/engine/game/g_18_zoo.rb [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+
+        @real_map = case @map
+                    when :map_b
+                      Engine::Game::G18ZOOMaps::MapB.new(@names)
+                    when :map_c
+                      Engine::Game::G18ZOOMaps::MapC.new(@names)
+                    when :map_d
+                      Engine::Game::G18ZOOMaps::MapD.new(@names)
+                    when :map_e
+                      Engine::Game::G18ZOOMaps::MapD.new(@names)
```
Comment (tobymao) [inline lib/engine/game/g_18_zoo.rb:91]: typos?

File: lib/engine/game/g_18_zoo.rb [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+      def end_game!
+        return if @finished
+
+        update_zootickets_value(4, 0)
+
+        super
+      end
+    end
+
+    module G18ZOOMaps
```
Comment (tobymao) [inline lib/engine/game/g_18_zoo.rb:601]: i don't think this should be done this way just have the maps and config as ruby constant hashes

File: lib/engine/step/buy_company.rb [hunk 5]
```diff
@@ -104,6 +99,16 @@ def round_state
       def setup
         @blocks = @opts[:blocks] || false
       end
+
+      def pay(entity, owner, price, company)
+        entity.spend(price, owner.nil? ? @game.bank : owner)
```
Comment (tobymao) [inline lib/engine/step/buy_company.rb:104]: owner ? game.bank : owner

File: lib/engine/step/g_18_zoo/additional_tracks_after_par.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+require_relative '../../action/lay_tile'
+
+module Engine
+  module Step
+    module G18ZOO
+      class AdditionalTracksAfterPar < Engine::Step::Track
+        ACTIONS = %w[lay_tile pass].freeze
+
+        def actions(entity)
+          return [] if @game.floated_corporation.nil?
```
Comment (tobymao) [inline lib/engine/step/g_18_zoo/additional_tracks_after_par.rb:14]: don't use nil change to unless .flotade-corporation

File: lib/engine/step/g_18_zoo/buy_sell_par_shares.rb [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        def actions(entity)
+          return [] unless entity == current_entity
+          return ['sell_shares'] if must_sell?(entity)
+
+          actions = []
+          actions << 'buy_shares' if can_buy_any?(entity)
+          actions << 'par' if can_ipo_any?(entity)
+          actions << 'buy_company' unless purchasable_unsold_companies.empty?
+          actions << 'sell_shares' if can_sell_any?(entity)
+          actions << 'sell_company' if can_sell_any_companies?(entity) && @game.floated_corporation.nil?
```
Comment (tobymao) [inline lib/engine/step/g_18_zoo/buy_sell_par_shares.rb:23]: don't call nil, do !floated_corporation

File: lib/engine/step/g_18_zoo/buy_sell_par_shares.rb [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+require_relative 'choose_power'
+
+module Engine
+  module Step
+    module G18ZOO
+      class BuySellParShares < BuySellParShares
+        include SellCompany
+        include ChoosePower
+
+        def actions(entity)
```
Comment (tobymao) [inline lib/engine/step/g_18_zoo/buy_sell_par_shares.rb:14]: why did you need to override this ?

File: lib/engine/step/g_18_zoo/buy_train.rb [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            corp.sms_hexes = @game.sms_hexes
+          end
+          @passed = true
+        end
+
+        def issuable_shares(entity)
+          return [] unless entity.corporation?
+          return [] unless entity.num_ipo_shares
+
+          []
```
Comment (tobymao) [inline lib/engine/step/g_18_zoo/buy_train.rb:54]: what is this

File: lib/engine/step/g_18_zoo/choose_power.rb [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+module Engine
+  module Step
+    module ChoosePower
+      def choice_name
+        'Use powers'
+      end
+
+      def choices
+        entity = current_entity
+        @choices = Hash.new { |h, k| h[k] = [] }
```
Comment (tobymao) [inline lib/engine/step/g_18_zoo/choose_power.rb:12]: why is this a default array when the values are strings?

File: lib/engine/step/g_18_zoo/choose_power.rb [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        @choices
+      end
+
+      def choice_available?(entity)
+        company_midas?(entity) || company_holiday?(entity) || company_whatsup?(entity) ||
+          company_greek_to_me?(entity)
+      end
+
+      def company_midas?(entity)
+        entity&.player? && entity.companies.any? { |c| c.sym == :MIDAS }
```
Comment (tobymao) [inline lib/engine/step/g_18_zoo/choose_power.rb:26]: do yo uhav ea refernece to midas? better to do entity.companiens.include?(midas)

File: lib/engine/step/g_18_zoo/track.rb [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        def lay_tile(action, _extra_cost: 0, _entity: nil, _spender: nil)
+          hex = action.hex
+          tile = action.tile
+
+          if %w[O M MM].include?(hex.location_name) &&
+            tile.color == :yellow &&
+            !tile.name.end_with?('(water)') && !tile.name.end_with?('(mountain)')
+            tile = case tile.name + '_' + hex.location_name
+                   when '7_O'
+                     Engine::Game::G18ZOO::TILE_W7.dup
```
Comment (tobymao) [inline lib/engine/step/g_18_zoo/track.rb:18]: what is this? why are you doing this? are you trying to make infinite tiles? you can just set that in the config if you are

File: lib/engine/step/g_18_zoo/special_track.rb [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+        def lay_tile(action, _extra_cost: 0, _entity: nil, _spender: nil)
+          hex = action.hex
+          tile = action.tile
+
+          if %w[O M MM].include?(hex.location_name) &&
+            tile.color == :yellow &&
+            !tile.name.end_with?('(water)') && !tile.name.end_with?('(mountain)')
+            tile = case tile.name + '_' + hex.location_name
+                   when '7_O'
+                     Engine::Game::G18ZOO::TILE_W7.dup
```
Comment (tobymao) [inline lib/engine/step/g_18_zoo/special_track.rb:18]: this looks copy and pasted...

**Comment by tobymao** [comment]
this is going to be very difficult to review, there's too much stuff. you have a ton of brand new steps and i can't tell what they are for or if they are just copy and pasted.

perhaps you could resubmit piecemeal so that we can digest the changes you are making

**Comment by carlorusso** [comment]
@tobymao maybe it's easier if I explain the reason of many of the changes I have done:
Every player get 3 free companies (zoo tickets) - they can sell them at any time during the SR.
If a player owns one of the four companies (MIDAS, ...) he/she can execute the ability at any time
to be able to do this I added the include to Steps SellCompany and ChoosePower inside FreeActionsOnSr and BuySellParShares

After the par the company must lay a home tile (if possible), and depending if par on green (9) or brown (12) they can lay directly a yellow (5,6,7), a green or a yellow (9), a brown, a green or a yellow (12)
After the lay home tile, if par at 9 or 12 they can lay additional free yellow tile (2 if par at 9, 4 if par at 12)
After the lay additional track you should be able to sell company or choose power if any (reason behind FreeActionsOnSr)

I have 6 different configuration (Map A...F) - every configuration share the train pool, the stock market, the companies - there are differences on the map, the corporations, cert limit (depends on map and player number)

As yellow track I have normal 7,8,9; but if you put one on the hex with location name O they should change into a town (and keep the terrain icon); if you put on the hex with location name M or MM they should keep the terrain icon; and also they should not be modifiable anymore.

There are 20 companies:
- 4 buyable in the draft, 4 for the 1st SR, 4 for the 2nd SR, 4 for the 3rd SR.
at the beginning of each SR you are able to know also which companies will be available in the next one (so I show them but you cannot buy)

So:
Auction:
- SimpleDraft - every player can buy just one of the companies available
SR:
- BuySellParShares - sell, buy, par shares; buy companies; sell companies; use abilities
- HomeTrackAfterPar - home track (yellow on par 5,6,7 - yellow,green on par 9 - yellow, green,brown on par 12)
- AdditionalTracksAfterPar - additional yellow track (2 if par 9, 4 if par 12)
- FreeActionsOnSr (block) - sell companies; use abilities (only if par, because in that case you exited from BuySellParShares step)
OR:
- SpecialTrack - needed to be able to add a track when using ability lay_track from company ANCIENT MAP
- BuyCompany - you can buy a company from players any time
- Track - normal track (but with special rule for water tiles)
- Token - normal token
- Route - every town will give money to the company
- Dividend - payout is based on threshold; if you get more than the threshold (depends on stock market) you get a value (depends on stock market) for each share you own - withhold is (revenue / 25)
- BuyCompany - you can buy a company from players any time
- BuyTrain - the first train you buy will move company on the right; if you buy the train opening the next phase you will move to the right;
- BuyCompany (block) - you can buy a company from players any time

For the OR I think I have to do many modification because I have to add abilities to the companies, so now it's just a draft

File: lib/engine/game/g_18_zoo.rb (hunk 1 above)
Comment (carlorusso) [reply lib/engine/game/g_18_zoo.rb:51]: after you par the first corporation you will be able to par only the two corporation around; then only the next in that direction;
example;
1) par PB
2) you can par only GI or PE
3) if you par GI then you can par only CR
3) if you par PE then you can par only LI
from 4) you can par only the following (or the previous), depending on the direction you followed when you par

File: lib/engine/step/g_18_zoo/track.rb (hunk 12 above)
Comment (carlorusso) [reply lib/engine/step/g_18_zoo/track.rb:18]: the tile 7,8,9 are changed into a new kind of tile (with town) when you put on the water terrain; so I don't have them as specific tile, but I'll change them on fly with a new one

File: lib/engine/step/g_18_zoo/special_track.rb (hunk 13 above)
Comment (carlorusso) [reply lib/engine/step/g_18_zoo/special_track.rb:18]: I need the same feature from track also here; I'm using a "super" at the end, and when I tried to put all of this inside a module I lost the "super" reference; I'll try to improve this anyway

**Comment by carlorusso** [comment]
Going to create smaller PR (I'll try to add any suggestion you already have done here)

**Comment by tobymao** [comment]
thanks, looking forward to getting this in!
